pkg/services/hcloud/server/server.go

```diff
@@ -90,62 +98,52 @@ func (s *Service) Reconcile(ctx context.Context) (res reconcile.Result, err erro
+func (s *Service) handleBootStateUnset(ctx context.Context) (reconcile.Result, error) {
+	m := s.scope.HCloudMachine
+	if m.Spec.ProviderID != nil && *m.Spec.ProviderID != "" {
+		// This machine seems to be an existing machine which was created before introducing
+		// ImageURL.
+
+		if !m.Status.Ready {
+			logger := ctrl.LoggerFrom(ctx)
+			logger.Info("fooooooooooooo") // TODO: remove this, when tests are fine.
+			panic("foooooooo")
+			return reconcile.Result{}, fmt.Errorf("Old machine, and status.Ready is false. Caph got updated while provisioning? We can't handle that.")
 		}
+		m.SetBootState(infrav1.HCloudBootStateOperatingSystemRunning)
+		return reconcile.Result{}, nil
 	}
 
-	s.scope.SetProviderID(server.ID)
-
-	// update HCloudMachineStatus
-	c := s.scope.HCloudMachine.Status.Conditions.DeepCopy()
-	sshKeys := s.scope.HCloudMachine.Status.SSHKeys
-	s.scope.HCloudMachine.Status = statusFromHCloudServer(server)
-	s.scope.SetRegion(failureDomain)
-	s.scope.HCloudMachine.Status.Conditions = c
-	s.scope.HCloudMachine.Status.SSHKeys = sshKeys
+	// This should not happen.
+	if m.Spec.ImageURL == "" && m.Spec.ImageName == "" {
+		return reconcile.Result{}, errors.New("imageName and imageURL are not given. Invalid Spec")
+	}
 
-	// validate labels
-	if err := validateLabels(server, s.createLabels()); err != nil {
-		err := fmt.Errorf("could not validate labels of HCloud server: %w", err)
-		s.scope.SetError(err.Error(), capierrors.CreateMachineError)
-		return res, nil
+	server, err := s.createServerFromImageNameOrURL(ctx)
+	if err != nil {
+		if errors.Is(err, errServerCreateNotPossible) {
+			return reconcile.Result{RequeueAfter: 5 * time.Minute}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("failed to create server: %w", err)
 	}
+	s.scope.SetProviderID(server.ID)
+	return reconcile.Result{RequeueAfter: requeueAfterCreateServer}, nil
+}
 
-	// analyze status of server
-	switch server.Status {
-	case hcloud.ServerStatusOff:
-		return s.handleServerStatusOff(ctx, server)
-	case hcloud.ServerStatusStarting:
-		// Requeue here so that server does not switch back and forth between off and starting.
-		// If we don't return here, the condition ServerAvailable would get marked as true in this
-		// case. However, if the server is stuck and does not power on, we should not mark the
-		// condition ServerAvailable as true to be able to remediate the server after a timeout.
-		conditions.MarkFalse(
-			s.scope.HCloudMachine,
-			infrav1.ServerAvailableCondition,
-			infrav1.ServerStartingReason,
-			clusterv1.ConditionSeverityInfo,
-			"server is starting",
-		)
-		return reconcile.Result{RequeueAfter: 1 * time.Minute}, nil
-	case hcloud.ServerStatusRunning: // do nothing
-	default:
-		// some temporary status
-		s.scope.SetReady(false)
-		return reconcile.Result{RequeueAfter: 10 * time.Second}, nil
+func (s *Service) handleBootToRealOS(ctx context.Context) (res reconcile.Result, reterr error) {
```

Comment:
Don't you have to use the switch on server.Status? I think you just removed it.. That was an important part

Reply:
@janiskemper I was aware of that. The code is near `// analyze status of server` now.